### PR TITLE
Refine the tools page side panel

### LINKED ISSUE #700: Refine the tools page side panel
On the tools page, we display the results of the daily GitHub crawl collected in `tools-data.json`. However, the side panels where one can select languages and authors are quite lengthy (see attached screenshot). To have a (searchable) combo box here instead would be a big improvement.

image

### PR DESCRIPTION
**Pull request by mattsmida**
Implement auto-complete comboboxes for languages and authors on the tools page's side panel.

Closes #700.

### CODE REVIEW

File: themes/bootstrap/layouts/_default/tools.html [hunk 1]
```diff
@@ -53,50 +53,32 @@
         </div>
         <div class="well">
           <fieldset>
```
Comment (lauft) [inline themes/bootstrap/layouts/_default/tools.html:55]: Let's merge the panels! 🤔

File: themes/bootstrap/layouts/_default/tools.html [hunk 2]
```diff
@@ -53,50 +53,32 @@
...
+            </legend>
+            <div class="combo js-multiselect" id="js-multiselect-languages">
+              <ul class="selected-options" id="languages-combo-selected"></ul>
+              <input type="text" name="language-input" id="languages-combo" placeholder="" role="combobox" class="combo-input" autocomplete="off"/>
+              <div class="combo-menu" role="listbox" id="languages-listbox"></div>
             </div>
           </fieldset>
         </div>
         <div class="well">
           <fieldset>
```
Comment (lauft) [inline themes/bootstrap/layouts/_default/tools.html:67]: Let's merge the panels! 🤔

File: themes/bootstrap/static/css/multiselect.css [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  text-align: left;
+  width: 100%;
+}
+
+.open .combo-input {
+  border-radius: 4px 4px 0 0;
+}
+
+.combo-input:focus {
+  border-color: #e60023;
```
Comment (lauft) [inline themes/bootstrap/static/css/multiselect.css:47]: Let's use a more blueish color for a focused combo box.

File: themes/bootstrap/layouts/_default/tools.html [hunk 4]
```diff
@@ -53,50 +53,32 @@
...
-        <p class="lead">
-          There
-          <span data-bind="visible: filteredTools().length != 1">are</span>
-          <span data-bind="visible: filteredTools().length == 1">is</span>
-          <span id="total_tools" data-bind="text: filteredTools().length">0</span>
-          <span data-bind="visible: filteredTools().length != 1">tools/extensions</span>
-          <span data-bind="visible: filteredTools().length == 1">tool/extension</span>
-          matching your search:
+        <p class="lead" id="search-result-message">
+          <!-- There are N tools/extensions matching your search: -->
```
Comment (lauft) [inline themes/bootstrap/layouts/_default/tools.html:81]: Out of curiosity: `There <span id="search-result-message"> matching your search:` is not possible?

File: themes/bootstrap/static/js/filter.js [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+          return a.name.localeCompare(b.name);
+      }
+      return b.rating - a.rating;
+  });
+  return sortedTools;
+}
+
+/** Given the tools data, return a sorted array of the languages
+ *  to be used for filtering.
+ */
```
Comment (lauft) [inline themes/bootstrap/static/js/filter.js:42]: Use [Google JavaScript code style](https://google.github.io/styleguide/jsguide.html#jsdoc-general-form) 
```suggestion
/** 
 * Given the tools data, return a sorted array of the languages to be used for filtering.
 */
```
In general, I would format all JSDoc like this (i.e. also those in line 16, 29, ...) and rather use _one sentence, one line_ than _hard wrap at X chars_...

File: themes/bootstrap/static/js/multiselect.js [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+  Next: 4,
+  Open: 5,
+  Previous: 6,
+  Select: 7,
+  Space: 8,
+  Type: 9 };
+
+
+// filter an array of options against an input string
+// returns an array of options that begin with the filter string, case-independent
```
Comment (lauft) [inline themes/bootstrap/static/js/multiselect.js:38]: Use JSDoc comment for function descriptions.

**Review by lauft — DISMISSED** [review]
Overall: I like the change, the search bar is much more compact now. 👍🏻 

I have left some comments with some ideas to further improve the layout.

File: themes/bootstrap/static/css/multiselect.css [hunk 7]
```diff
@@ -0,0 +1,151 @@
+/** Multiselect component adapted from
+    https://webdevtrick.com/html-combobox-with-javascript/
+  */
+
```
Comment (lauft) [inline themes/bootstrap/static/css/multiselect.css:4]: The combo box seems to have the same height as the search box. Would it be possible to make them a bit narrower? Could we display a placeholder text like `select language` (or just `select...`, if there is not enough space)?

File: themes/bootstrap/static/js/filter.js [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+          </strong>&ensp;
+          <span style="display: ${tool.dormant ? 'visible' : 'none'};" class="tooltip">
+            <em class="bi bi-moon-stars"></em>
+            <span class="tooltip-text">project is dormant</span>
+          </span>
+          <span style="display: ${tool.archived ? 'visible' : 'none'}" class="tooltip">
+            <em class="bi bi-archive"></em>
+            <span class="tooltip-text">project has been archived</span>
+          </span><br>
+          <span style="display: ${tool.owner.length > 0 ? 'visible' : 'none'}>
```
Comment (lauft) [inline themes/bootstrap/static/js/filter.js:142]: Add missing double colon:
```suggestion
          <span style="display: ${tool.owner.length > 0 ? 'visible' : 'none'}">
```

File: themes/bootstrap/static/js/filter.js [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+            <small>
+              by <em title="Owners">${tool.owner.join(', ')}</em>
+            </small>
+          </span>
+        </p>
+        <p>
+          <span>Rating:
+            <span title="Rating">${tool.rating}</span>
+            &ensp;<em class="bi bi-star"></em>
```
Comment (lauft) [inline themes/bootstrap/static/js/filter.js:151]: Remove linebreak to reduce whitespace between rating and star icon
```suggestion
            <span title="Rating">${tool.rating}</span>&ensp;<em class="bi bi-star"></em>
```

File: themes/bootstrap/layouts/_default/tools.html (hunk 4 above)
Comment (mattsmida) [reply themes/bootstrap/layouts/_default/tools.html:81]: Decided to initialize this with a `loading` message, and then update the whole message on state change. Reason: with the proposed `<span>`, the message's length changes in a jarring way on search.

**Comment by mattsmida** [comment]
Still pending: filters for category.

**Comment by mattsmida** [comment]
In addition to addressing the comments in this thread, this PR now includes a tools page that can be filtered based on category. For ti.net, such filtering is turned off by default -- the plan is to deploy it on the timewarrior site with minimal changes.

**Comment by mattsmida** [comment]
I have converted this PR to a draft pending the resolution of [the similar issue on timewarrior.org](https://github.com/GothenburgBitFactory/tw.org/pull/866). Let's finish that one before returning to this PR.

**Pull request merged**